core/src/main/java/io/questdb/griffin/engine/functions/groupby/ApproxPercentileDoubleGroupByFunction.java

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.engine.functions.*;
+import io.questdb.std.Numbers;
+import io.questdb.std.histogram.org.HdrHistogram.DoubleHistogram;
+import io.questdb.std.histogram.org.HdrHistogram.Histogram;
+
+public class ApproxPercentileDoubleGroupByFunction extends DoubleFunction implements GroupByFunction, BinaryFunction {
+    // specifies the precision for the recorded values (between 0 and 5).
+    // trade-off between memory usage and accuracy.
+    private final int numberOfSignificantValueDigits = 3;
+    private final Function left;
+    private final Function right;
+    private DoubleHistogram histogram;
```

Comment:
Hi @puzpuzpuz I have a question. I noticed in other group by functions, [CountDistinctLongGroupByFunction](https://github.com/questdb/questdb/blob/aec8b347ba5f5402b6d7470b64d57ec9fe989dd5/core/src/main/java/io/questdb/griffin/engine/functions/groupby/CountDistinctLongGroupByFunction.java#L39) for example, it has a `valueIndex`, `setIndex`, and a list of `sets` as attributes.

Is the purpose of this for handling concurrent requests? So like each request is given an index to prevent its result being overwritten, or what is it for? 

Initially I thought each request instantiates a new instance of the function, so I was a bit confused when i saw those index attributes.

Reply:
Hi Kevin,

`CountDistinctLongGroupByFunction` as well as other `GroupByFunction` functions operate on GROUP BY group, i.e. when you run `SELECT key, sum(value) FROM tab;` each distinct `key` value means a separate group (row in the result set) and has to have its own `sum(value)` aggregate.

`CountDistinctLongGroupByFunction` has a few methods that are invoked by the upper-level code:
* `computeFirst` - this method is invoked when a new group (key) is first met returned from the base cursor
* `computeNext` - this one is invoked when a known key is met

Considering this, the function does its bookkeeping in the `sets` field (this is a list of hash sets for all known groups) and the `setIndex` field. `setIndex` is nothing more but the last position in the `sets` list, used for extending the list in append-only manner. `valueIndex` is the index of the value "column" (or a field) in the `FastMap` used to store the key-value pairs accumulating the aggregate data. So, the map allows `CountDistinctLongGroupByFunction` to find the index in the `sets` list when it handles a known group.

All of the above is single-threaded, no concurrency is involved.

Hopefully, this makes sense.

Reply:
Oh right, I totally missed that, it makes more sense now. Thanks for the explanation Andrei!